### ISSUE TITLE
Crash on startup with particular scenes

### ISSUE THREAD

**Issue by atobenzo** [opentoonz/opentoonz#3292]
**Describe the bug**
This has now happened twice: at some point, my scene can no longer be opened because it gives a warning message: "There were problems loading the scene. Some files may be missing", and then crashes with a segmentation fault on Linux. The same thing happens if I use the Windows version under WINE.  After the first time I made sure that I was regularly saving backups, but they all fail in the same way.

**To Reproduce**
List steps to reproduce the behavior.
1. Open my scene file 

**Expected behavior**
The scene opens

**Device Information:**
 - OS: Ubuntu 19.10
 - OpenToonz Version: either 1.4 or a self-compiled master, same issue
 - GPU: GTX 1050 Ti Mobile or Intel integrated graphics, same issue
 - GPU Driver Version: [e.g. 440.00]

[gadget3.tnz.txt](https://github.com/opentoonz/opentoonz/files/4646734/gadget3.tnz.txt)

**Comment by manongjohn** [comment]
I think this is related to #3168 

I corrected your file, I think.  See if this works: 
[gadget3.tnz.txt](https://github.com/opentoonz/opentoonz/files/4646954/gadget3.tnz.txt)

**Comment by atobenzo** [comment]
You're a lifesaver, thank you! So if it happens again I'll just remove the weird keyframe nodes. Let me know if I can help to debug to find out why the issue happens.

**Comment by manongjohn** [comment]
Look for `<n>` and `<>` nodes in keyframe data. That usually is what causes the crashes.

I believe it's something with how the keyframes are either being cut/pasted, how interpolations are being modified or a little bit of both.  Debugging this will likely be difficult since it doesn't show up until you save. You'd have to make a change, save and look for this bad data. That or make a note of what you were doing with keyframes/interpolations.

Out of curiosity, How were you manipulating keyframes and interpolations in this scene? Using fuction editor? using xsheet? cutting/pasting? something else?

**Comment by atobenzo** [comment]
Yeah, thinking about it both times I had been cut/pasting keyframes... I'll let you know if I come up with a minimal change that causes this.

**Comment by AnyhowStep** [comment]
@manongjohn This looks suspicious,

https://github.com/opentoonz/opentoonz/blob/4334db09bc5349c9294c2bd79ba12703f0a779bb/toonz/sources/toonz/keyframedata.cpp#L175-L177

The TL;DR is,

I believe the intended line is,
```
newKey.m_type = nextKey.m_type
```

Right now, it is,
```
newKey.m_type = nextKey.m_prevType
```

-----

I believe `m_prevType` is of type `TDoubleKeyframe::Type` and isn't initialized in the constructor initialization list.
So, it should have a default value of `0`, or `None`.
So, `newKey.m_type = nextKey.m_prevType` should set `m_type` to `None`, in most cases.

-----

The comment for that snippet seems to also suggest that line is the cause of the bug,
```
// If new key was added between existing keys, sync new key to the next
// key's previous interpolation
```

I was, indeed, copy-pasting keys in between existing keys.

-----

I'm not 100% sure about this, though. I hope it's somewhat useful.

**Comment by manongjohn** [comment]
@AnyhowStep Thanks for trying to look into this. It should **not** be `newKey.m_type = nextKey.m_type`.

Consider the following:

m_type is the value of the interpolation starting at that key.  It can change from key to key.  
m_prevType should indicate what the previous key's interpolation was.

Let's say you have 1 key and you create a key to the **right** of it, the creation of right key dictates what interpolation type is used (preference setting) from the left (existing) to the right (new) .  As such, the left key's m_type is updated to the new interpolation which should match the right key's m_prevType.  The right key's m_type should be blank or none because there is no key after it.

Let's do the reverse.  Let's start with 1 key and you create a key to the **left** of it.  The left key dictates what the interpolation type is used (preference setting) from the left (new) to the right (existing).  So the left key's m_type is copied to the right key's m_prevType (it was blank/none previously).  The left key's m_prevType should be blank or none because there is no key before it.

Now let's say we have 2 keys and we want to create a key between them.  The middle key's interpolation (m_type) should be dictated by either the right key's m_prevType value or left key's m_type value to keep the original interpolation between left and right keys.  I chose to update the middle key's m_type from the right key's m_prevType.

**Comment by manongjohn** [comment]
I should add that if the m_prevType value is not set correctly, then copy/pasting with the current logic only proliferates the bad data.

Looking over the pseudo logic, it could be because of how copy/paste of keys is done. Code might be different when handling pasting of keys.

If a different piece of code handles pasted key insertion, need to review that logic.

If it also uses this logic, I think I see a flaw in it. The logic for the middle key insertion currently assumes the new key is brand new and doesn't handle if it's a pasted key which could have a different interpolation type than what is already there. So pasting a copied key in the middle would force an interpolation change per current logic. If it is a pasted key, i think it should actually update the right key's m_prevType to match the middle key. 

In addition to that, the m_prevType key on the middle key my not be updated to the left key's m_type which also needs to be done on a paste. So if the copied key was the 1st key then the m_prevType may still be blank when pasted in the middle. Paste another key before that and we would then create these bad scenarios.

Needs confirmation and further review.

**Comment by RodneyBaker** [comment]
@atobenzo Any update on the status of this report?
This issue is likely resolved in nightly releases (as of May 2020 with other subsequently related fixes) which will also be included in the v1.5 release.
